[InstCombine] Optimize unneeded float to int cast when icmp

trokhymchuk opened the pull request:
Fixes #154242 

The following folds are supported:

* `icmp sgt (fptosi %x), <negative> -> fcmp ogt %x, <negative>`: https://alive2.llvm.org/ce/z/etWGrN
* `icmp sgt (fptosi %x), <non-negative> -> fcmp oge %x, <non-negative + 1>`: https://alive2.llvm.org/ce/z/sdZTgm
* `icmp slt (fptosi %x), <positive> -> fcmp olt %x, <positive>`: https://alive2.llvm.org/ce/z/d78YAh
* `icmp slt (fptosi %x), <non-positive> -> fcmp ole %x, <non-positive - 1>:` https://alive2.llvm.org/ce/z/kwQv2U

trokhymchuk committed:
[InstCombint] add icmp <cond> (fptosi %x) -> fcmp <cond> %y optimization tests

Add optimization test reducing unneded float-to-int cast when
comparing numbers:
* icmp sgt (fptosi %x), <negative> -> fcmp ogt %x, <negative>
* icmp sgt (fptosi %x), <non-negative> -> fcmp oge %x, <non-negative + 1>
* icmp slt (fptosi %x), <positive> -> fcmp olt %x, <positive>
* icmp slt (fptosi %x), <non-positive> -> fcmp ole %x, <non-positive - 1>

trokhymchuk committed:
[InstCombine] add icmp <cond> (fptosi %x) -> fcmp <cond> %y optimization implementation

Add optimization test reducing unneded float-to-int cast when
comparing numbers:
* icmp sgt (fptosi %x), <negative> -> fcmp ogt %x, <negative>
* icmp sgt (fptosi %x), <non-negative> -> fcmp oge %x, <non-negative + 1>
* icmp slt (fptosi %x), <positive> -> fcmp olt %x, <positive>
* icmp slt (fptosi %x), <non-positive> -> fcmp ole %x, <non-positive - 1>

trokhymchuk commented on llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:
```diff
@@ -7611,6 +7613,208 @@ Instruction *InstCombinerImpl::foldICmpCommutative(CmpPredicate Pred,
+/// \return result constant
+static Constant *castIntegralConstantToFloat(const Constant *C, Type *FPType,
+                                             int Addend) {
+  auto getInnerType = [FPType]() -> Type * {
+    if (FPType->isFloatingPointTy())
+      return FPType;
+
+    if (!FPType->isVectorTy())
+      return nullptr;
+
+    return llvm::cast<llvm::VectorType>(FPType)->getElementType();
+  };
```
lambda is not needed, jsut calculate type once

trokhymchuk commented on llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:
```diff
@@ -7611,6 +7613,208 @@ Instruction *InstCombinerImpl::foldICmpCommutative(CmpPredicate Pred,
+  if (!getInnerType() || !getInnerType()->isFloatingPointTy())
+    return nullptr;
+
+  if (const ConstantInt *CI = dyn_cast<ConstantInt>(C)) {
+    if (auto ConverstionResult = castCIToAPF(CI, FPType, Addend);
+        ConverstionResult.first == APFloat::opOK)
+      return ConstantFP::get(FPType, ConverstionResult.second);
+
+    return nullptr; // failed to convert int-to-float
+  }
```
need to remove nested IFs

trokhymchuk commented:
Looks like I also need advice: current implementation relies upon the fact that constant is **always** second argument to the `icmp` and I was unable to thwart the canonicalization, because when I used constant in the 'thwart instruction' I just got constant folded result and 'thwarting' did not change the order of operands, and I am unsure how to thwart correctly or should I support reverse order at all.

nikic commented:
> Looks like I also need advice: current implementation relies upon the fact that constant is **always** second argument to the `icmp` and I was unable to thwart the canonicalization, because when I used constant in the 'thwart instruction' I just got constant folded result and 'thwarting' did not change the order of operands, and I am unsure how to thwart correctly or should I support reverse order at all.

Constants for icmp are always on the right, there is no need to test them on the left. Also the thwarting bit in https://llvm.org/docs/InstCombineContributorGuide.html#id8 is actually outdated (no longer needed). I'll update the docs.

trokhymchuk commented on llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:
```diff
@@ -7611,6 +7613,208 @@ Instruction *InstCombinerImpl::foldICmpCommutative(CmpPredicate Pred,
+/// \return pair {cast status, APFloat result}
+static std::pair<APFloatBase::opStatus, APFloat>
+castCIToAPF(const ConstantInt *CI, const Type *FPType, int Addend) {
+  APFloat FVal{FPType->isFloatTy() ? APFloat::IEEEsingle()
```
thank you, used `FPType->getFltSemantics` function and added double and half float tests
I was unable to add vector of half floats as `@llvm.convert.from.fp16` refused to convert

trokhymchuk commented on llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:
```diff
@@ -7611,6 +7613,208 @@ Instruction *InstCombinerImpl::foldICmpCommutative(CmpPredicate Pred,
+/// \param Addend addend to add before casting
+///
+/// \return result constant
+static Constant *castIntegralConstantToFloat(const Constant *C, Type *FPType,
```
indeed, I updated the code to use those functions (and removed `castIntegralConstantToFloat` completly), thanks for the suggestion

trokhymchuk commented on llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:
```diff
@@ -7611,6 +7613,208 @@ Instruction *InstCombinerImpl::foldICmpCommutative(CmpPredicate Pred,
+
+  Value *FloatOp = FToSI->getOperand(0);
+  if (!FloatOp)
+    return nullptr;
```
updated the code, pls take a look

trokhymchuk commented on llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:
```diff
@@ -7611,6 +7613,208 @@ Instruction *InstCombinerImpl::foldICmpCommutative(CmpPredicate Pred,
+///
+/// \return folded instruction or nullptr, if failed to combine instructions
+static Instruction *foldICmpFToSIToFCmp(ICmpInst &ICmp, InstCombiner &IC) {
+  if (ICmp.getPredicate() != ICmpInst::ICMP_SGT &&
```
I used matcher instead of manually casting and checking, pls take a look

about non-splat vector: looks like after I updated the implementation to use `match(C, m_CheckedInt(CheckSign));` there would be little to no simplification, if non-splat vector would not be supported, what do you think?

trokhymchuk commented on llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:
```diff
@@ -7611,6 +7613,208 @@ Instruction *InstCombinerImpl::foldICmpCommutative(CmpPredicate Pred,
+  Value *New;
+  // fcmp FCmpPredicate %arg C
+  New = B.CreateFCmp(FCmpPredicate, FloatOp, FCmpConstant);
+  return IC.replaceInstUsesWith(ICmp, New);
```
thanks, updated

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:
```diff
@@ -7611,6 +7614,168 @@ Instruction *InstCombinerImpl::foldICmpCommutative(CmpPredicate Pred,
+    return nullptr;
+
+  return ConstantDataVector::getSplat(CDV->getNumElements(), CIAddend);
+}
```
This whole function should be `ConstantInt::getSigned(C->getType(), Addend)`.

trokhymchuk replied:
thanks for comment, I did not expect that was so easy, thank you

I updated the code, pls, take a look

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:
```diff
@@ -7611,6 +7614,168 @@ Instruction *InstCombinerImpl::foldICmpCommutative(CmpPredicate Pred,
+                                             int Addend, const DataLayout &DL) {
+  Type *InnerType{getInnerType(FPType, &Type::isFloatingPointTy)};
+  if (!InnerType || !InnerType->isFloatingPointTy())
+    return nullptr;
```
`isFPOrFPVectorTy()`

nikic commented on llvm/test/Transforms/InstCombine/icmp.ll:
```diff
@@ -6054,3 +6054,255 @@ define i1 @icmp_samesign_logical_or(i32 %In) {
+; CHECK-NEXT:    [[V2:%.*]] = fcmp olt float [[V0]], -1.000000e+00
+; CHECK-NEXT:    ret i1 [[V2]]
+;
+  %v0 = call float @llvm.convert.from.fp16(i16 %arg0)
```
Do not use these intrinsics, this should just be `fptosi half %v0 to i32`.

trokhymchuk replied:
sorry, looks like I was confused with the [IR reference](https://llvm.org/docs/LangRef.html#llvm-convert-to-fp16-intrinsic) :( 

 I updated the code to use `half` type and added vector tests forthe half-floats

nikic reviewed (commented):
(Drive-by comments only)

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:
```diff
@@ -31,6 +33,7 @@
 #include "llvm/Support/KnownBits.h"
 #include "llvm/Transforms/InstCombine/InstCombiner.h"
 #include <bitset>
+#include <utility>
```
Unused header.

trokhymchuk replied:
removed the header, thanks

dtcxzyw commented:
@zyw-bot mfuzz

dtcxzyw reviewed (changes requested):
Miscompilation reproducer: https://alive2.llvm.org/ce/z/3hmRxH

trokhymchuk commented on llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:
```diff
@@ -7611,6 +7614,117 @@ Instruction *InstCombinerImpl::foldICmpCommutative(CmpPredicate Pred,
+  if (!FPType->isFPOrFPVectorTy())
+    return nullptr;
+
+  Constant *CWithAddend{ConstantFoldBinaryOpOperands(
```
updated the code, pls take a look

trokhymchuk commented:
> Miscompilation reproducer: https://alive2.llvm.org/ce/z/3hmRxH

yep, my bad, messed with comparators and did not check the optimization result ones, fixed now, @dtcxzyw pls take a look

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:
```diff
@@ -7611,6 +7613,117 @@ Instruction *InstCombinerImpl::foldICmpCommutative(CmpPredicate Pred,
+      return Value.isNegative();
+    case SignType::NonNegative:
+      return Value.isNonNegative();
+    default:
```
/home/gha/actions-runner/_work/llvm-project/llvm-project/llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7641:5: error: default label in switch which covers all enumeration values [-Werror,-Wcovered-switch-default]

nikic replied:
This helper is generally unnecessary, you can directly use `m_NonNegative()` etc instead.

dtcxzyw replied:
It is used to support non-spalt vector constants. See https://github.com/llvm/llvm-project/pull/155501?notification_referrer_id=NT_kwDOAO7OmbQxODQ5MDk1OTY0OToxNTY1MDQ1Nw#discussion_r2311989751.

nikic replied:
The m_NonNegative() etc matchers that do not accept an argument also match non-splat vectors.

dtcxzyw commented on llvm/test/Transforms/InstCombine/icmp.ll:
```diff
@@ -6054,3 +6054,273 @@ define i1 @icmp_samesign_logical_or(i32 %In) {
   ret i1 %V
 }
+
+; https://alive2.llvm.org/ce/z/XtQS6H
```
Please add a case where the integer value cannot be represented with the fp type(e.g., `icmp sgt i32 (fptosi half %x), -65535`).

trokhymchuk replied:
added tests with `-65535` and `65535` that cannot be represented as half, thank you

dtcxzyw commented:
@trokhymchuk Can you please provide a generalized proof? See https://llvm.org/docs/InstCombineContributorGuide.html#proofs.

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:
```diff
@@ -7611,6 +7613,117 @@ Instruction *InstCombinerImpl::foldICmpCommutative(CmpPredicate Pred,
+    return nullptr;
+
+  FCmpInst::Predicate FCmpPredicate;
+  Constant *FCmpConstant{};
```
```suggestion
  Constant *FCmpConstant;
```

trokhymchuk replied:
same as [below](https://github.com/llvm/llvm-project/pull/155501#discussion_r2328648280)

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:
```diff
@@ -7611,6 +7613,117 @@ Instruction *InstCombinerImpl::foldICmpCommutative(CmpPredicate Pred,
+    return nullptr;
+
+  if (Pred != ICmpInst::ICMP_SGT && Pred != ICmpInst::ICMP_SLT)
+    return nullptr;
```
You can use `default: return nullptr` instead of the separate check.

trokhymchuk replied:
I also had this thought (using pointer as boolean marker does not look pretty to me also) the change would also require adding additional `else` to ifs: if non-splat vector with values `[1, -1]` it is neither positive, non-positive, negative nor non-negative and thus neither if branch would be executed. `nullptr` works for that case, but default would not. Additionally, `FCmpConstant` checked for nullptr works when integer is failed to convert to fp number.

trokhymchuk commented:
> @trokhymchuk Can you please provide a generalized proof? See https://llvm.org/docs/InstCombineContributorGuide.html#proofs.

generalized proofs:

* `icmp sgt (fptosi %x), <negative> -> fcmp ogt %x, <negative>`: https://alive2.llvm.org/ce/z/isTzu2
* `icmp sgt (fptosi %x), <non-negative> -> fcmp oge %x, <non-negative + 1>`: https://alive2.llvm.org/ce/z/sdZTgm
* `icmp slt (fptosi %x), <positive> -> fcmp olt %x, <positive>`: https://alive2.llvm.org/ce/z/d78YAh
* `icmp slt (fptosi %x), <non-positive> -> fcmp ole %x, <non-positive - 1>:` https://alive2.llvm.org/ce/z/kwQv2U

trokhymchuk commented on llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:
```diff
@@ -7611,6 +7613,117 @@ Instruction *InstCombinerImpl::foldICmpCommutative(CmpPredicate Pred,
+/// \return result constant
+static Constant *castIntegralConstantToFloat(Constant *C, Type *FPType,
+                                             int Addend, const DataLayout &DL) {
+  if (!FPType->isFPOrFPVectorTy())
```
replaced with assert,pls take a look

trokhymchuk commented on llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:
```diff
@@ -7611,6 +7613,117 @@ Instruction *InstCombinerImpl::foldICmpCommutative(CmpPredicate Pred,
+  if (!FPType->isFPOrFPVectorTy())
+    return nullptr;
+
+  Constant *CWithAddend = ConstantFoldBinaryOpOperands(
```
added check for nullptr, pls take a look

trokhymchuk committed:
[InstCombine] add icmp <cond> (fptosi %x) -> fcmp <cond> %y fix regression tests

Fix regression in the llvm/test/Transforms/PGOProfile/chr.ll after
optimizing fptosi cast in icmp.

trokhymchuk commented:
While running regression tests found problem with `Transforms/PGOProfile/chr.ll` test, looks like current optimization applies there also, fixed in the separate commit (b65aa2db7b0a47e32711adeb81962e72bd4f43e6) (pls comment if the commit must be squashed with previous)

nikic commented:
> * `icmp sgt (fptosi %x), <non-negative> -> fcmp oge %x, <non-negative + 1>`: https://alive2.llvm.org/ce/z/sdZTgm
> * `icmp slt (fptosi %x), <non-positive> -> fcmp ole %x, <non-positive - 1>:` https://alive2.llvm.org/ce/z/kwQv2U

It looks like these two have an additional pre-condition on the min/max value, but it doesn't seem to be checked in your implementation?
